appenders/python/pulse/metrics.py

```diff
@@ -0,0 +1,167 @@
+        :param capacity: The maximum number of metrics to buffer before
+                                flushing. If no value is provided, each metric
+                                will be submitted one at a time. Defaults to
+                                1000.
+        :param threadCount: Number of threads to handle post requests.
+        :type threadCount: int
+        :param logger: :class:`logging.Logger` object for debug logging. If
+                       none is provided, a StreamHandler will be created to log
+                       to sys.stdout. It is recommended that you provide a
+                       logger if you plan to use more than one instance of this
+                       class.
+        :type logger: logging.Logger
+        :rtype: MetricWriter
+        """
+        PulseBatcher.__init__(self, endpoint, capacity, threadCount, logger)
+        self.table = table
+        # Cleanup when Python terminates
+        atexit.register(self.close)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def gaugeTimestamp(self, key, tag, value, timestamp,
+                       frmt="%Y-%m-%dT%H:%M:%S.%f"):
+        """
+        Gauge metric at the provided timestamp.
+
+        :param key: Key or ID of the gauge
+        :type key: int or str
+        :param tag: Tag or name of the gauge
+        :type tag: str
+        :param value: Value of the gauge.
+        :type value: float
+        :param timestamp: Timestamp at which to gauge the metric.
+                          If timezone is not provided, UTC is assumed.
+        :type timestamp: str or :class:`datetime.datetime`
+        :param frmt: Required if a string timestamp is provided
+                     Uses the C strftime() function, see strftime(3)
+                     documentation.
+        :type frmt: str
+        """
+        self.logger.debug("Constructing metric data")
+        data = dict()
+        data["id"] = key
+        data["metric"] = tag
+        data["value"] = value
+        data["ts"] = self.unixTimeMicros(timestamp, frmt)
+
+        self.handle(data)
+
+    def gauge(self, key, tag, value):
+        """
+        Gauge metric at the current timestamp.
+
+        :param key: Key or ID of the gauge
+        :type key: int or str
+        :param tag: Tag or name of the gauge
+        :type tag: str
+        :param value: Value of the gauge.
+        :type value: float
+        :rtype: requests.Response
+        """
+        self.gaugeTimestamp(key, tag, value, datetime.utcnow())
+
+    def flush(self):
+        """
+        Flush items from buffer and clear buffer.
+        """
+        # Overrides PulseBatcher.flush
+        self.buffer = {"table_name": self.table, "payload": self.buffer}
+        self.queue.put(dict(self.buffer), block=False)
+        self.buffer = list()
+
+    @staticmethod
+    def unixTimeMicros(timestamp, frmt="%Y-%m-%dT%H:%M:%S.%f"):
```

Comment:
Pretty incredible what it takes to get microseconds!

Reply:
agreed